dhall/src/Dhall/Import.hs

```diff
@@ -463,88 +465,376 @@ localToPath prefix file_ = liftIO $ do
+
+    let semanticHash = Crypto.Hash.hash bytes
+    let dependencies = map graph imports
+    let graph = Tree.Node import_ dependencies
+    return (SemanticImport {..})
+
+
+-- Fetch source code directly from disk/network
+fetchFresh :: ImportStack -> ResolvedImportType -> StateT Status IO Text
+fetchFresh _ (ResolvedLocal prefix file) = liftIO $ do
+    path <- localToPath prefix file
+    exists <- Directory.doesFileExist path
+    if exists
+        then Data.Text.IO.readFile path
+        else throwMissingImport (MissingFile path)
+
+fetchFresh stack (ResolvedRemote url mHeaders) = do
+    let encodeHeader (key, value) =
+            (Data.CaseInsensitive.mk (Data.Text.Encoding.encodeUtf8 key)
+            , Data.Text.Encoding.encodeUtf8 value)
+    text <- fetchFromHttpUrl stack url (fmap (map encodeHeader) mHeaders)
+    return text
+
+fetchFresh _ (ResolvedEnv env) = liftIO $ do
+    x <- System.Environment.lookupEnv (Text.unpack env)
+    case x of
+        Just string -> do
+            return (Text.pack string)
+        Nothing -> do
+                throwMissingImport (MissingEnvironmentVariable env)
+
+fetchFresh _ ResolvedMissing = liftIO $ throwM (MissingImports [])
+
+-- Fetch encoded normal form from "semantic cache"
+fetchFromSemanticCache :: SemanticHash -> IO (Maybe Data.ByteString.ByteString)
+fetchFromSemanticCache expectedHash = Maybe.runMaybeT $ do
+    cacheFile <- getCacheFile expectedHash
+    True <- liftIO (Directory.doesFileExist cacheFile)
+    liftIO (Data.ByteString.readFile cacheFile)
+
+-- Fetch encoded normal form from "semi-semantic cache"
+fetchFromSemisemanticCache :: SemisemanticHash -> IO (Maybe Data.ByteString.ByteString)
+fetchFromSemisemanticCache semisemanticHash = Maybe.runMaybeT $ do
+        cacheFile <- getCacheFile semisemanticHash  -- TODO: use different cache directory
+        True <- liftIO (Directory.doesFileExist cacheFile)
+        liftIO (Data.ByteString.readFile cacheFile)
+
+-- write to "semantic cache"
+writeToSemanticCache :: SemanticHash -> Data.ByteString.ByteString -> IO ()
+writeToSemanticCache semanticHash bytes = do
     _ <- Maybe.runMaybeT $ do
-        Just expectedHash  <- return hash
-        cacheFile          <- getCacheFile expectedHash
-
-        _ <- Dhall.Core.throws (Dhall.TypeCheck.typeWith _startingContext expression)
+        cacheFile <- getCacheFile semanticHash
+        liftIO (Data.ByteString.writeFile cacheFile bytes)
+    return ()
 
-        let normalizedExpression =
-                Dhall.Core.alphaNormalize
-                    (Dhall.Core.normalizeWith
-                        _normalizer
-                        expression
-                    )
+-- write to "semi-semantic cache"
+writeToSemisemanticCache :: SemisemanticHash -> Data.ByteString.ByteString -> IO ()
+writeToSemisemanticCache semisemanticHash bytes = do
+    _ <- Maybe.runMaybeT $ do
+        cacheFile <- getCacheFile semisemanticHash  -- TODO: different cache location
```

Comment:
If you're looking for a cache location, I'd suggest trying, in order, `${XDG_CACHE_HOME}/dhall-haskell` and `~/.cache/dhall-haskell`, since it's currently a Haskell-specific optimization.